### PR TITLE
[Concurrency] Emit warning suggesting async function

### PR DESCRIPTION
**Pull request by etcwilde**
This patch adds a diagnostic to recommend using the async function
instead of using the completion handler version. The diagnostic is only
emitted when the call is made in an async context and the
completion-handler version of the function is attributed with the
@completionHandleyAsync attribute.

Resolves: rdar://75221638

### CODE REVIEW

**Comment by etcwilde** [comment]
@swift-ci please smoke test

**Review by ktoso — APPROVED** [review]
Nice one!

Would also offering a a fixitReplace be a bit too hard?

**Comment by etcwilde** [comment]
> Nice one!
Thanks!

> Would also offering a a fixitReplace be a bit too hard?
Yeah, Doug and I discussed what this would take and agreed that it probably should be a refactoring step and not a fix-it. Things like parameter re-ordering and pulling the code inside the completion handler out into the calling context make it a bit more challenging than a fix-it.

**Comment by etcwilde** [comment]
-- forgot to run `git clang-format` before pushing. Nothing semantic changed!

@swift-ci please smoke test

**Comment by bnbarham** [comment]
> > Nice one!
> 
> Thanks!
> 
> > Would also offering a a fixitReplace be a bit too hard?
> 
> Yeah, Doug and I discussed what this would take and agreed that it probably should be a refactoring step and not a fix-it. Things like parameter re-ordering and pulling the code inside the completion handler out into the calling context make it a bit more challenging than a fix-it.

It is definitely a more involved fixit than anything we have now, but I was planning on adding it. The way the refactoring works is by replacing the whole call, so that should work fine with a fixit. We'd have to pull out the implementation from refactoring and make it a little more generic, but that shouldn't be *too* bad.

Note that the refactoring doesn't handle parameter re-ordering at the moment, but it wouldn't hurt to make it more robust regardless (it should check it at the very least).

Happy to leave the fixit out of this PR though :)

File: test/attr/attr_completionhandlerasync.swift [hunk 1]
```diff
@@ -137,3 +142,21 @@ func matchingAsyncFunc(value: String) async {} // expected-note{{'matchingAsyncF
 // expected-error@+1:25{{ambiguous '@completionHandlerAsync' async function 'matchingAsyncFunc(value:)'}}
 @completionHandlerAsync("matchingAsyncFunc(value:)")
 func typecheckFunc9(handler: @escaping () -> Void) {}
+
+// Suggest using async alternative function in async context
+
```
Comment (bnbarham) [inline test/attr/attr_completionhandlerasync.swift:147]: Might be worth adding some tests for the closure case above. Another case is to check we don't output a warning for references of either top level/instance functions.

**Review by bnbarham — COMMENTED** [review]
Thanks Evan! I had forgotten to add a radar for this, so I'm glad you remembered :)

File: lib/Sema/TypeCheckConcurrency.h [hunk 2]
```diff
@@ -221,6 +221,13 @@ enum class ConcurrentValueCheck {
 bool checkConcurrentValueConformance(
     ProtocolConformance *conformance, ConcurrentValueCheck check);
 
+/// Check that the async overload
```
[the remark this answers is not in the document]
Comment (etcwilde) [reply lib/Sema/TypeCheckConcurrency.h:224]: No, I just forgot to finish my thought. :) Actually trying to describe what the function does.

File: include/swift/AST/DiagnosticsSema.def [hunk 3]
```diff
@@ -3370,6 +3370,9 @@ ERROR(attr_completion_handler_async_ambiguous_function,none,
 ERROR(attr_completion_handler_async_no_suitable_function,none,
       "no corresponding async function named %0", (DeclNameRef))
 
+WARNING(warn_use_async_alternative,none,
+        "consider using asynchronous alternative function",())
```
[the remark this answers is not in the document]
Comment (etcwilde) [reply include/swift/AST/DiagnosticsSema.def:3374]: Are the refactorings something that is Xcode-specific, or can they be used on other platforms/editors? If the refactors only work on the one platform/IDE, it'll look pretty weird for anyone working on other platforms.

[the remark this answers is not in the document]
Comment (bnbarham) [reply include/swift/AST/DiagnosticsSema.def:3374]: Eh, you're right - would look weird in a compiler message from eg. the command line.

**Comment by etcwilde** [comment]
Sigh, so it just occurred to me that someone might want to do something like:

```swift
let asyncClosure = { () async -> () in }

@completionHandlerAsync("asyncClosure")
let compHandlerClosure = { (_ handler: @escaping () -> ()) -> () in }

@main struct Main {
  static func main() async {
    compHandlerClosure { print("yuup") }
  }
}
```

Checking this will take a bit more work, so I'll take care of that in a separate PR if that's alright.

**Comment by bnbarham** [comment]
> Sigh, so it just occurred to me that someone might want to do something like:
> 
> ```swift
> let asyncClosure = { () async -> () in }
> 
> @completionHandlerAsync("asyncClosure")
> let compHandlerClosure = { (_ handler: @escaping () -> ()) -> () in }
> 
> @main struct Main {
>   static func main() async {
>     compHandlerClosure { print("yuup") }
>   }
> }
> ```
> 
> Checking this will take a bit more work, so I'll take care of that in a separate PR if that's alright.

Does `@completionHandlerAsync` even work in that case at the moment? I thought we restricted it to functions?

**Comment by etcwilde** [comment]
> > Sigh, so it just occurred to me that someone might want to do something like:
> > ```swift
> > let asyncClosure = { () async -> () in }
> > 
> > @completionHandlerAsync("asyncClosure")
> > let compHandlerClosure = { (_ handler: @escaping () -> ()) -> () in }
> > 
> > @main struct Main {
> >   static func main() async {
> >     compHandlerClosure { print("yuup") }
> >   }
> > }
> > ```
> > 
> > 
> > Checking this will take a bit more work, so I'll take care of that in a separate PR if that's alright.
> 
> Does `@completionHandlerAsync` even work in that case at the moment? I thought we restricted it to functions?

No, it doesn't right now. I think this starts getting into property-wrapper syntactic territory, so maybe just leave it? I don't know, but it seems like something _someone_ might try to do. I'm going to leave it for now though and pick it back up if it becomes a problem.

**Comment by bnbarham** [comment]
> No, it doesn't right now. I think this starts getting into property-wrapper syntactic territory, so maybe just leave it? I don't know, but it seems like something _someone_ might try to do. I'm going to leave it for now though and pick it back up if it becomes a problem.

Sounds good to me!

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 4]
```diff
@@ -3060,3 +3060,38 @@ NormalProtocolConformance *GetImplicitConcurrentValueRequest::evaluate(
 
   return conformance;
 }
+
+namespace {
+class UsageChecker : public ASTWalker {
```
Comment (DougGregor) [inline lib/Sema/TypeCheckConcurrency.cpp:3065]: I recommend folding this into `MiscDiagnostics.cpp`, which has a whole bunch of these structural checks.

**Comment by etcwilde** [comment]
@swift-ci please smoke test

**Comment by etcwilde** [comment]
@swift-ci please smoke test

**Review by bnbarham — APPROVED** [review]
Just those couple test additions, but otherwise LGTM

**Comment by etcwilde** [comment]
@swift-ci please smoke test

File: test/attr/attr_completionhandlerasync.swift [hunk 5]
```diff
@@ -137,3 +143,61 @@ func matchingAsyncFunc(value: String) async {} // expected-note{{'matchingAsyncF
...
+    // expected-warning@+1{{consider using asynchronous alternative function}}
+    compAsyncHandlerFunc() { [self] () async -> () in
+      // expected-warning@+1{{consider using asynchronous alternative function}}
+      compAsyncHandlerFunc() { [self] () async -> () in
+        // expected-warning@+1{{consider using asynchronous alternative function}}
+        compHandlerFunc() { print("foo") }
+      }
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (etcwilde) [reply test/attr/attr_completionhandlerasync.swift:233]: Okay, I think this covers the various test cases for where the functions can be declared and called.

[the remark this answers is not in the document]
Comment (bnbarham) [reply test/attr/attr_completionhandlerasync.swift:233]: Thanks!

**Pull request merged**